### PR TITLE
Linear algebra changes

### PR DESCRIPTION
**Pull request by klevzoff**
Changes in LAI accumulated while working on other things. There is a fair bit of refactoring in the implementation of custom hypre kernels. I split this out into a separate PR to make reviewing easier, and also because several of us are working on linear algebra concurrently and it's good to synchronize.

Some new methods/kernels added are only used in multiscale branch (still WIP). I will place comments as appropriate. It would take more work for me to split changes further, so I'm lumping all LAI changes together here.

All tests pass (on CPU) with no diffs.

### CODE REVIEW

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    case 7: return 5;
+    case 8:
+    case 13:
+    case 14: return 4;
+    case 18: return 1;
+    default: return -1;
+  }
+}
+
+HYPRE_Int RelaxationSetup( HYPRE_Solver solver,
```
[the remark this answers is not in the document]
Comment (rrsettgast) [reply src/coreComponents/linearAlgebra/interfaces/hypre/HypreUtils.cpp:96]: maybe document the reason for the `reinterpret_cast` in the code comments and provide some comments about maintaining the cast.

File: src/coreComponents/linearAlgebra/interfaces/trilinos/TrilinosPreconditioner.cpp [hunk 2]
```diff
@@ -97,7 +97,8 @@ string getMLSmootherType( LinearSolverParameters::AMG::SmootherType const & valu
     { LinearSolverParameters::AMG::SmootherType::default_, "Chebyshev" },
     { LinearSolverParameters::AMG::SmootherType::jacobi, "Jacobi" },
     { LinearSolverParameters::AMG::SmootherType::l1jacobi, "Jacobi" },
-    { LinearSolverParameters::AMG::SmootherType::gs, "Gauss-Seidel" },
+    { LinearSolverParameters::AMG::SmootherType::fgs, "Gauss-Seidel" },
+    { LinearSolverParameters::AMG::SmootherType::bgs, "Gauss-Seidel" },
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/interfaces/trilinos/TrilinosPreconditioner.cpp:101]: Maybe have a different string for these?

File: src/coreComponents/linearAlgebra/solvers/GmresSolver.cpp [hunk 3]
```diff
@@ -70,42 +67,39 @@ void ApplyGivensRotation( real64 const c, real64 const s, real64 & dx, real64 &
   dx = temp;
 }
 
-void Backsolve( localIndex const k,
+void Backsolve( integer const k,
                 arraySlice2d< real64 const, MatrixLayout::COL_MAJOR > const & H,
                 arraySlice1d< real64 > const & g )
 {
-  for( localIndex j = k - 1; j >= 0; --j )
+  for( integer j = k - 1; j >= 0; --j )
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/solvers/GmresSolver.cpp:74]: this is not a `localIndex`?

File: src/coreComponents/mainInterface/initialization.cpp [hunk 4]
```diff
@@ -117,6 +118,7 @@ std::unique_ptr< CommandLineOptions > parseCommandLineOptions( int argc, char *
     { OUTPUTDIR, 0, "o", "output", Arg::nonEmpty, "\t-o, --output, \t Directory to put the output files" },
     { TIMERS, 0, "t", "timers", Arg::nonEmpty, "\t-t, --timers, \t String specifying the type of timer output." },
     { SUPPRESS_MOVE_LOGGING, 0, "", "suppress-move-logging", Arg::None, "\t--suppress-move-logging \t Suppress logging of host-device data migration" },
+    { PAUSE_FOR, 0, "p", "pause-for", Arg::numeric, "\t-p, --pause-for, \t Pause geosx for a given number of seconds before starting execution" },
```
[the remark this answers is not in the document]
Comment (rrsettgast) [reply src/coreComponents/mainInterface/initialization.cpp:121]: Totalview is not a blessing these days!

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/linearAlgebra/interfaces/MatrixBase.hpp [hunk 5]
```diff
@@ -237,7 +257,7 @@ class MatrixBase : public virtual LinearOperator< VECTOR >
     }
 
     createWithLocalSize( localMatrix.numRows(),
-                         localMatrix.numRows(),
+                         numLocalColumns,
```
[the remark this answers is not in the document]
Comment (klevzoff) [reply src/coreComponents/linearAlgebra/interfaces/MatrixBase.hpp:260]: I think we ended up not touching it precisely for the reason that I had already done it in this PR

File: src/coreComponents/linearAlgebra/solvers/GmresSolver.cpp (hunk 3 above)
Comment (klevzoff) [reply src/coreComponents/linearAlgebra/solvers/GmresSolver.cpp:74]: No, here `k` (and by extension `j` and `i`) varies between 0 and the size of Krylov subspace (i.e. max number of restart iterations), which is on the order of ~100. I've recently been replacing many uses of `localIndex` that I find incorrect, i.e. whenever a quantity does not scale with the size of local mesh partition.

**Pull request merged**